### ISSUE TITLE
Potential PHP 8.0 incompatibility?

### ISSUE THREAD

**Issue by paulgibbs** [cedaro/satispress#189]
There seems to be a PHP 8 incompatibility, as seen on WP Engine, on PHP 8.0. PHP 7.4 (the only other option they support) works as expected.

Error log output is here https://gist.github.com/paulgibbs/c39a87669114f3b0a5e1343769ddd2e6

**Comment by sdrib** [comment]
Can confirm we have the same issue. PHP 8.1

`Uncaught TypeError: preg_replace(): Argument #3 ($subject) must be of type array|string, bool given` 

The `send_header` in`RequestHandler.php` receives a bool at one point: the value is false.

Evaluating `$value` in the function, only allowing array|string(or before) stops the fatal error

**Comment by IanRandall** [comment]
I'm seeing a similar error on PHP 8.0 (fine on 7.4)

PHP message: PHP Fatal error:  Uncaught TypeError: preg_replace(): Argument #3 ($subject) must be of type array|string, bool given in /wordpress/wp-content/plugins/satispress/src/Provider/RequestHandler.php:149"

WARNING: [pool www] child 291 said into stderr: "PHP message: PHP Fatal error:  Uncaught TypeError: preg_replace(): Argument #3 ($subject) must be of type array|string, bool given in /efs/wordpress/wp-content/plugins/satispress/src/Provider/RequestHandler.php:149"
WARNING: [pool www] child 291 said into stderr: "Stack trace:"
WARNING: [pool www] child 291 said into stderr: "#0 /efs/wordpress/wp-content/plugins/satispress/src/Provider/RequestHandler.php(149): preg_replace('/\\s+/', ' ', false)"
WARNING: [pool www] child 291 said into stderr: "#1 /efs/wordpress/wp-content/plugins/satispress/src/Provider/RequestHandler.php(162): SatisPress\Provider\RequestHandler->send_header('Last-Modified', false)"
WARNING: [pool www] child 291 said into stderr: "#2 /efs/wordpress/wp-content/plugins/satispress/src/Provider/RequestHandler.php(102): SatisPress\Provider\RequestHandler->send_headers(Array)"
WARNING: [pool www] child 291 said into stderr: "#3 /efs/wordpress/wp-includes/class-wp-hook.php(307): SatisPress\Provider\RequestHandler->dispatch(Object(WP))"
WARNING: [pool www] child 291 said into stderr: "#4 /efs/wordpress/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters(NULL, Array)"
[error] 192#192: *533 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught TypeError: preg_replace(): Argument #3 ($subject) must be of type array|string, bool given in //wordpress/wp-content/plugins/satispress/src/Provider/RequestHandler.php:149
WARNING: [pool www] child 291 said into stderr: "#5 /efs/wordpress/wp-includes/plugin.php(524): WP_Hook->do_action(Array)"
WARNING: [pool www] child 291 said into stderr: "#6 /efs/wordpress/wp-includes/class-wp.php(398): do_action_ref_array('parse_request', Array)"
WARNING: [pool www] child 291 said into stderr: "#7 /..."

Stack trace:
#0 /wordpress/wp-content/plugins/satispress/src/Provider/RequestHandler.php(149): preg_replace('/\\s+/', ' ', false)
#1 /wordpress/wp-content/plugins/satispress/src/Provider/RequestHandler.php(162): SatisPress\Provider\RequestHandler->send_header('Last-Modified', false)
#2 /wordpress/wp-content/plugins/satispress/src/Provider/RequestHandler.php(102): SatisPress\Provider\RequestHandler->send_headers(Array)
#3 /wordpress/wp-includes/class-wp-hook.php(307): SatisPress\Provider\RequestHandler->dispatch(Object(WP))
#4 /wordpress/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters(NULL, Array)
#5 /wordpress/wp-includes/plugin.php(524): WP_Hook->do_action(Array)
#6 /wordpress/wp-includes/class-wp.php(398): do_action_ref_array('parse_request', Array)

Response: HTTP Code: 500 Bytes Sent: 3011 Time: 1.540

**Comment by bradyvercher** [comment]
@paulgibbs @sdrib @IanRandall There are several related issues pointing to the same line. I believe this should have been taken care of in [this commit](https://github.com/cedaro/satispress/commit/803a353501480567c3c3e39b39b01d261ca39fce), though. Can y'all verify you're running the latest version of SatisPress?

`RequestHandler::send_header()` is pretty much a direct copy of `WP_REST_Server::send_header()`, so any issues are likely due to an incorrect header being sent at some point.

**Comment by IanRandall** [comment]
Will take a look, thank you @bradyvercher

**Comment by james0r** [comment]
Any news on this? Seeing the composer require fails for PHP version 
```php
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - cedaro/satispress[v1.0.0, ..., v1.0.2] require php ^7 -> your php version (8.1.0) does not satisfy that requirement.
    - cedaro/satispress v1.0.3 requires composer/installers ^1.0 -> found composer/installers[v1.0.0, ..., v1.12.0] but the package is fixed to v2.2.0 (lock file version) by a partial update and that version does not match. Make sure you list it as an argument for the update command.
    - cedaro/satispress 1.0.4 requires psr/container ^1.0 -> found psr/container[1.0.0, 1.1.0, 1.1.1, 1.1.2] but the package is fixed to 2.0.2 (lock file version) by a partial update and that version does not match. Make sure you list it as an argument for the update command.
    - Root composer.json requires cedaro/satispress ^1.0 -> satisfiable by cedaro/satispress[v1.0.0, ..., 1.0.4].
    ```
and installing via .zip gives a `SatisPress is missing required dependencies.` upon activation.

**Comment by chrillep** [comment]
maybe use rector?
https://github.com/rectorphp/rector

**Comment by bradyvercher** [comment]
I believe all PHP incompatibilities were addressed with the release of 2.0. Please open another issue if there are any other problems.
